source/module_base/test/complexarray_test.cpp

```diff
@@ -84,14 +107,78 @@ class ComplexArray_test : public testing::Test
 
 };
 
-TEST_F(ComplexArray_test,operator_equal)
+
+TEST(ComplexArray, construction1_test)
+{
+    ModuleBase::ComplexArray a(1,2,3,4);
+    ASSERT_EQ(a.getSize(),24);
+    ASSERT_EQ(a.getBound1(),1);
+    ASSERT_EQ(a.getBound2(),2);
+    ASSERT_EQ(a.getBound3(),3);
+    ASSERT_EQ(a.getBound4(),4);
+    /*
+    ASSERT_DEATH(ModuleBase::ComplexArray a(0,2,3,4),"");
+    ASSERT_DEATH(ModuleBase::ComplexArray a(1,0,3,4),"");
+    ASSERT_DEATH(ModuleBase::ComplexArray a(1,2,0,4),"");
+    ASSERT_DEATH(ModuleBase::ComplexArray a(1,2,3,0),"");
+    */
+}
+
+TEST(ComplexArray, construction2_test)
+{
+    ModuleBase::ComplexArray a(1,2,3,4);
+    ModuleBase::ComplexArray b(a);
+    ASSERT_EQ(b.getSize(),24);
+    ASSERT_EQ(b.getBound1(),1);
+    ASSERT_EQ(b.getBound2(),2);
+    ASSERT_EQ(b.getBound3(),3);
+    ASSERT_EQ(b.getBound4(),4);
+       
+    ASSERT_EQ(a.getSize(),24);
+    ASSERT_EQ(a.getBound1(),1);
+    ASSERT_EQ(a.getBound2(),2);
+    ASSERT_EQ(a.getBound3(),3);
+    ASSERT_EQ(a.getBound4(),4);
+
+}
+
+TEST(ComplexArray, construction3_test)
+{
+    ModuleBase::ComplexArray b(ModuleBase::ComplexArray(1,2,3,4));
+    ASSERT_EQ(b.getSize(),24);
+    ASSERT_EQ(b.getBound1(),1);
+    ASSERT_EQ(b.getBound2(),2);
+    ASSERT_EQ(b.getBound3(),3);
+    ASSERT_EQ(b.getBound4(),4);    
+}
+
+
+TEST_F(ComplexArray_test,operator_equal_complex)
 {   
-    c2 = a2; //c2 is just constructed as a CompleArray, but not assigned by any value
-    b2 = a2; //b2 is constructed as a ComplexArray, and is also assigned by a complex
-    EXPECT_EQ(c2,a2);
-    EXPECT_EQ(b2,a2);
-    EXPECT_NE(c2,a4);
-    EXPECT_DEATH(c2=a4,"");
+    for(int i = 0;i<a2.getSize();++i)
+    {EXPECT_EQ(a2.ptr[i],com1);}
+}
+
+
+TEST_F(ComplexArray_test,operator_equal_ComplexArray)
+{   
+    ModuleBase::ComplexArray b = a2;
+    for ( int i = 0;i<a2.getSize();++i)
+    {
+        EXPECT_EQ(b.ptr[i],a2.ptr[i]);
```

Comment:
Why not using `EXPECT_DOUBLE_EQ` if contents in `ComplexArray` class are stored in double?

Reply:
It is a complex value, and can not used in EXPECT_DOUBLE_EQ

complexarray_test.cpp(456): error: no instance of function template "testing::internal::CmpHelperFloatingPointEQ" matches the argument list
            argument types are: (const char [10], const char [10], std::complex<double>, std::complex<double>)
          EXPECT_DOUBLE_EQ(c2.ptr[i],com1*com2);

Reply:
> It is a complex value, and can not used in EXPECT_DOUBLE_EQ
> 
> complexarray_test.cpp(456): error: no instance of function template "testing::internal::CmpHelperFloatingPointEQ" matches the argument list argument types are: (const char [10], const char [10], std::complex, std::complex) EXPECT_DOUBLE_EQ(c2.ptr[i],com1*com2);

I see. So, we have to compare the real part and imag part separately with `EXPECT_DOUBLE_EQ`?